JEP.md

```diff
@@ -0,0 +1,293 @@
+
+**Jupyter components:**
+
+* Jupyter server
+* JupyterLab
+* JupyterHub
+* Jupyter Classic
+
+## Detailed Explanation
+
+### Event Schema
+
+Each event is published conforming a given event schema and this is the contract followed by each step in the event lifecycle. These are referred to by all publishers regardless of whether the event comes from the server or the browser.
+
+Example event schema
+
+```json
+{
+    "name": "org.jupyter.kernel_lifecycle_event",
+    "version": 1,
+    "title": "Event emitted when a kernel lifecycle state changes",
+    "description": "Records each invocation of any command in JupyterLab",
+    "type": "object",
+    "args": {
+        "kernel_name": {
+            "type": "string",
+            "description": "Name of the kernel"
+        },
+        "state": {
+            "type": "string",
+            "description": "The new state of the kernel"
+        }
+    }
+}
+```
+
+Schema validation is done in the core telemetry framework that routes events from publishers to sinks, so that each event sink does not need to implement schema validation.
+
+**Open Questions**:
+
+* How do organizations deploy custom schemas?
+* Where are the schemas for public events stored?
+
+### Event Sinks
+
+Event sinks are the backends where events published to.  Event sinks can be configured from the browser as well as the server. Custom sinks can be implemented by extending the interface for the given Jupyter component. 
+See [Python interface](#python-event-sink-interface) and the [JupyterLab interface](#jupyterlab-event-sink-interface)
+
+### Jupyter Server
+
+#### Python publisher library
+
+The Python event publisher library provides extension developers and other internal components an interface to record events. This is agnostic of the event sink back-end and the other deployment configurations.
+
+```python
+import jupyter.telemetry as telemetry
+
+telemetry.record_event(
+    name='org.jupyter.kernel_lifecycle_event',
+    schema_version='1',
+    args={
+        'kernel_name': 'python3',
+        'state': 'restarted'
+    }
+)
+```
+
+#### Python Event Sink Interface
+
+The event sink interface
+
+```python
+class MyCustomEventSink(JupyterEventSink):
+
+    def consume_event(event_data):
+        # Business logic to publish event into the backend.
+```
+
+#### Configuration
+
+A setting on the Jupyter server allows operators to configure various options of the event sinks such as:
+
+* The event sink implementations
+* The list of event names to whitelist or blacklist
+
+```bash
+jupyter notebook \
+    --NotebookApp.telemetry_event_sinks=mypackage.sinks.MyCustomEventSink,mypackage.sinks.AnotherCustomEventSink \
+    --NotebookApp.blacklisted_events='org.jupyter.someevent' \
+```
+
+#### Core event router
+
+The implementation of the `telemetry.record_event` method which handles
+
+* Schema validation
+* Adds event metadata such as `creationTime`
```

Comment:
We should define a way to provide an 'event capsule' that is automatic fields we put on every event. They should not clash with any properties defined in the schema right now or in the future. In https://github.com/jupyterhub/jupyterhub/pull/2542 we use the Python `__<name>__` syntax for it, and forbid schema properties to have any fields starting with `__`. This also lets us (and admins) add arbitrary capsule properties without worrying about clashing with schema properties.

Reply:
That is what I meant by `Adds event metadata such as `creationTime`.  I don't know if capsule is the best terminology and if everyone will understand.

Are you saying that there is a need for letting operators define this metadata/capsule?

Reply:
Ah, sure. The terminology doesn't matter - just that it's automatically added.

> Are you saying that there is a need for letting operators define this metadata/capsule?

More that the eventlogging mechanism / library should be able to add more metadata fields in the future if necessary without clashing with any potential schemas that have been defined in the various uses.

Reply:
> should be able to add more metadata fields in the future if necessary without clashing with any potential schemas that have been defined in the various uses.

Cool - I added this as a note in the doc.